sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/SortPushDown.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader;
+
+/**
+ * A mix in interface for `DataSourceV2Reader`. Users can implement this interface to pre-sort
+ * the data and avoid sorting at Spark side.
+ */
+public interface SortPushDown {
+  /**
+   * Returns true if the implementation can handle this sorting requirement and save a sort
+   * operation at Spark side.
```

Comment:
Like the discussion on hash support, I would rather see this the other way around, where the data source can reports its underlying sort order. Maybe there are some sources that can perform a sort before passing the data to Spark, but its hard to know when it should. I think the more useful case is when the data is already sorted.

Reply:
ok I agree, shall we distinguish global sort and per-partition sort?

Reply:
I'd leave this out of the first iteration and add it later, but some thoughts: specify the preferred sort orders, and the data source can return the sortedness.

Reply:
Per-partition Sort would be great, We usually can't handle a Global sort but there are some per-partition sorts (By partition here I mean SQL Style Partitioning) we can handle. 

```SELECT pkey, RANK() OVER (PARTITION BY pkey ORDER BY clustCol) FROM ks.tab```

Like in this case if pkey is the C* partition key then there is a natural ordering we can take advantage of when returning values.